### PR TITLE
[SPARK-34233][SQL] FIX NPE for char padding in binary comparison

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

we need to check whether the `lit` is null  before calling `numChars`

### Why are the changes needed?

fix an obvious NPE bug

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

new tests

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 1]
```diff
@@ -146,6 +146,22 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
     }
   }
 
+  test("char/varchar with null value for partitioned columns") {
```
Comment (yaooqinn) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:149]: this test will fail temporary due to other bugs being fixed

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu @HyukjinKwon @dongjoon-hyun, can you check this?

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:149]: Got it. BTW, please add `SPARK-34233: ` prefix.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 2]
```diff
@@ -146,6 +146,22 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
...
+    Seq("CHAR(5)", "VARCHAR(5)").foreach { typ =>
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:150]: `typ` -> `type`.

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 3]
```diff
@@ -146,6 +146,22 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
...
+      withTable("t") {
+        sql(s"CREATE TABLE t(i STRING, c $typ) USING $format PARTITIONED BY (c)")
+        sql("INSERT INTO t VALUES ('1', null)")
+        checkPlainResult(spark.table("t"), typ, null)
+        sql("INSERT OVERWRITE t VALUES ('1', null)")
+        checkPlainResult(spark.table("t"), typ, null)
+        sql("INSERT OVERWRITE t PARTITION (c=null) VALUES ('1')")
+        checkPlainResult(spark.table("t"), typ, null)
+        sql(s"ALTER TABLE t DROP PARTITION(c=null)")
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:159]: `s"` -> `"`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 4]
```diff
@@ -3957,13 +3957,16 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
...
-              val targetLen = (length +: literalCharLengths).max
-              Some(i.copy(
-                value = addPadding(attr, length, targetLen),
-                list = list.zip(literalCharLengths).map {
-                  case (lit, charLength) => addPadding(lit, charLength, targetLen)
-                }))
+              val (nulls, literalChars) =
+                list.map(_.eval().asInstanceOf[UTF8String]).partition(_ == null)
+              val literalCharLengths = literalChars.map(_.numChars())
+              val targetLen = literalCharLengths.fold(length)(_ max _)
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3963]: Why do we change this line? It looks functionally identical to me.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 5]
```diff
@@ -3957,13 +3957,16 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
...
+
+              val newAttr = addPadding(attr, length, targetLen)
+              val newList = list.zip(literalCharLengths).map {
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3966]: Is this change required?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 6]
```diff
@@ -3957,13 +3957,16 @@ object ApplyCharTypePadding extends Rule[LogicalPlan] {
...
+                case (lit, charLength) => addPadding(lit, charLength, targetLen)
+              } ++ nulls.map(Literal.create(_, StringType))
+              Some(i.copy(value = newAttr, list = newList))
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: If you don't mind, shall we keep the change minimal? For example, it seems that we don't need to change the following lines.
```scala
Some(i.copy(
  value = addPadding(attr, length, targetLen),
```

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: +1, let's minimize the diff

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:150]: Hi @dongjoon-hyun, 
this is the 2nd time I see your comment on `type`/`typ` :). 
The `type` is a keyword in scala and usually, we use `typ` or `xxxType` instead.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:3969]: OK, got it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala (hunk 4 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala]: reverted

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala]: thanks

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala [hunk 7]
```diff
@@ -443,7 +460,29 @@ trait CharVarcharTestSuite extends QueryTest with SQLTestUtils {
...
+  private def testNullConditions(df: DataFrame, conditions: Seq[String]): Unit = {
+    conditions.foreach { cond =>
+      checkAnswer(df.selectExpr(cond), Row(null))
+    }
+  }
+
+  test("SPARK-34233: char type comparison with null values") {
+    val conditions = Seq("c1 = null", "c1 IN ('e', null)", "c1 IN (null)")
+    withTable("t") {
+      sql(s"CREATE TABLE t(c1 CHAR(2), c2 CHAR(5)) USING $format")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:477]: why do we need the `c2` column while all the conditions only refer to `c1`?

**Review by cloud-fan — APPROVED** [review]

**Review by HyukjinKwon — APPROVED** [review]

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 1 above)
Comment (yaooqinn) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala]: let's wait for #31339

File: sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala (hunk 2 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/CharVarcharTestSuite.scala:150]: Oh.. My bad. Thank you for saying that, @yaooqinn .

**Review by dongjoon-hyun — COMMENTED** [review]
The newly added test case seems to fail in `DSv2`. Could you take a look at this?
```
DSV2CharVarcharTestSuite.SPARK-34233: char/varchar with null value for partitioned columns
org.apache.spark.sql.AnalysisException: Cannot translate expression to source filter: null
```

**Comment by yaooqinn** [comment]
> The newly added test case seems to fail in `DSv2`. Could you take a look at this?
> 
> ```
> DSV2CharVarcharTestSuite.SPARK-34233: char/varchar with null value for partitioned columns
> org.apache.spark.sql.AnalysisException: Cannot translate expression to source filter: null
> ```

As mentioned https://github.com/apache/spark/pull/31336#discussion_r564196108, this is another bug being fixed. For test completeness in this PR, I'd wait for you guys to verify and merge that first as the solution there is clear to me. Or I can move this test the one of its subclasses temporary. WDYT?

**Comment by cloud-fan** [comment]
GA passed, merging to master/3.1!

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM.

**Comment by dongjoon-hyun** [comment]
Hi, this breaks `branch-3.1` UTs. Please take a look at it.

**Comment by dongjoon-hyun** [comment]
Since this only fails at branch-3.1, I'll revert this first at `branch-3.1` only in order to recover the branch.
- https://github.com/apache/spark/runs/1780772467
- https://amplab.cs.berkeley.edu/jenkins/view/Spark%20QA%20Test/job/spark-branch-3.1-test-maven-hadoop-3.2-scala-2.13/195/testReport/junit/org.apache.spark.sql/DSV2CharVarcharTestSuite/SPARK_34233__char_varchar_with_null_value_for_partitioned_columns/
- Manual
```
$ build/sbt "sql/testOnly *.DSV2CharVarcharTestSuite -- -z SPARK-34233"
...
[info] DSV2CharVarcharTestSuite:
20:01:55.767 WARN org.apache.hadoop.util.NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
[info] - SPARK-34233: char/varchar with null value for partitioned columns *** FAILED *** (4 seconds, 202 milliseconds)
[info]   Results do not match for query:
[info]   Timezone: sun.util.calendar.ZoneInfo[id="America/Los_Angeles",offset=-28800000,dstSavings=3600000,useDaylight=true,transitions=185,lastRule=java.util.SimpleTimeZone[id=America/Los_Angeles,offset=-28800000,dstSavings=3600000,useDaylight=true,startYear=0,startMode=3,startMonth=2,startDay=8,startDayOfWeek=1,startTime=7200000,startTimeMode=0,endMode=3,endMonth=10,endDay=1,endDayOfWeek=1,endTime=7200000,endTimeMode=0]]
[info]   Timezone Env: 
[info]   
[info]   == Parsed Logical Plan ==
[info]   'UnresolvedRelation [t], [], false
[info]   
[info]   == Analyzed Logical Plan ==
[info]   i: string, c: string
[info]   SubqueryAlias testcat.t
[info]   +- RelationV2[i#54, c#55] testcat.t
[info]   
[info]   == Optimized Logical Plan ==
[info]   RelationV2[i#54, c#55] testcat.t
[info]   
[info]   == Physical Plan ==
[info]   *(1) Project [i#54, c#55]
[info]   +- BatchScan[i#54, c#55] class org.apache.spark.sql.connector.InMemoryTable$InMemoryBatchScan
[info]   
[info]   == Results ==
[info]   
[info]   == Results ==
[info]   !== Correct Answer - 0 ==   == Spark Answer - 1 ==
[info]   !struct<>                   struct<i:string,c:string>
[info]   !                           [1,null] (QueryTest.scala:243)
```

Sorry guys. Please make a revised PR to branch-3.1, @yaooqinn .

**Comment by yaooqinn** [comment]
Thanks, @dongjoon-hyun I will look into this today

**Comment by dongjoon-hyun** [comment]
Thank you. Take your time. According to @HyukjinKwon 's email, the upcoming RC cut is next Monday.
